libs/pika/async_mpi/tests/unit/pool_creation.cpp

```diff
@@ -0,0 +1,152 @@
+//  Copyright (c) 2024 ETH Zurich
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#include <pika/config/compiler_specific.hpp>
+#include <pika/debugging/print.hpp>
+#include <pika/exception.hpp>
+#include <pika/execution.hpp>
+#include <pika/init.hpp>
+#include <pika/mpi.hpp>
+#include <pika/testing.hpp>
+
+#include <atomic>
+#include <cstdlib>
+#include <mpi.h>
+#include <string>
+#include <utility>
+#include <vector>
+
+/// The purpose of this test is to ensure that a manually created pool,
+/// can still be used even though by default, we can now let pika create an mpi pool for us
+/// This test does one auto create and two manual creates to check if all works without fail
+/// note that it is expected that this test will print warnings when a pool is created
+/// but not actually needed by the completion mode
+
+static const std::string random_name1 = "abcd12345qwerty";
+static const std::string random_name2 = "ta-daa-500";
+
+namespace ex = pika::execution::experimental;
+namespace mpi = pika::mpi::experimental;
+namespace tt = pika::this_thread::experimental;
+
+// -----------------------------------------------------------------
+int pika_main(const std::string& pool_name)
+{
+    int size, rank;
+    MPI_Comm comm = MPI_COMM_WORLD;
+    MPI_Comm_size(comm, &size);
+    MPI_Comm_rank(comm, &rank);
+
+    MPI_Datatype datatype = MPI_INT;
+
+    {
+        // Register polling on our custom pool
+        mpi::enable_polling enable_polling(mpi::exception_mode::install_handler, pool_name);
+        // Success path
+        {
+            // MPI function pointer
+            int data = 0, count = 1;
+            if (rank == 0) { data = 42; }
+            auto s = mpi::transform_mpi(ex::just(&data, count, datatype, 0, comm), MPI_Ibcast);
+            tt::sync_wait(PIKA_MOVE(s));
+            PIKA_TEST_EQ(data, 42);
+        }
+
+        // Operator| overload
+        {
+            // MPI function pointer
+            int data = 0, count = 1;
+            if (rank == 0) { data = 42; }
+            tt::sync_wait(
+                ex::just(&data, count, datatype, 0, comm) | mpi::transform_mpi(MPI_Ibcast));
+            PIKA_TEST_EQ(data, 42);
+        }
```

Comment:
I don't think you this test needs to use both the regular function call and the pipe version here, since it's tested elsewhere already. Feel free to only use only whichever version you find more natural to use in each case.